### PR TITLE
feat: Add helper to programmatically update React components outside of a React context

### PR DESCRIPTION
**Pull request by jonathonherbert**
## What does this change?

Adds a helper function that allows us to programmatically update React components outside of their React context. This is useful when our CMS state informs our elements about what they should be renderering.

See the component itself and its tests for details on usage – hopefully, it's self-documenting, and any feedback to improve either API or docs would be great.

This is one approach to solving this problem, but others exist – definitely approach with an open mind! There are a few advantages to doing things in this way: 
- it doesn't extend the library API in solving what is essentially a UI problem
- it provides an imperative way of updating state (useful for consumers) that only rerenders the parts of the application that depend on that state, so it should be reasonably efficient

## How to test

- Does the new component make sense, conceptually and in use? Is there anything we could do to improve usability?
- Do the tests pass, and cover the functionality of the component?
- Does the new button, `Randomise image role options`, affect the image role options in the Image element as expected?

### CODE REVIEW

**Comment by SHession** [comment]
It doesn't look like the checks have run on this PR, I'm not sure why.

File: src/elements/image/ImageElement.tsx
Comment (SHession) [inline src/elements/image/ImageElement.tsx:112]: We might want to be consistent with out external facing API by calling `updateRoleOptions` -> `updateAdditionalRoleOptions`.

File: src/elements/image/ImageElementForm.tsx
Comment (SHession) [inline src/elements/image/ImageElementForm.tsx:77]: Still on the fence about this being a separate component. It might look a bit tidier if it was split out, but I acknowledge the number of props it would require.

File: src/elements/image/ImageElementForm.tsx
Comment (SHession) [inline src/elements/image/ImageElementForm.tsx:97]: Not an ideal solution, but should do the job!

File: src/plugin/types/Element.ts
Comment (SHession) [inline src/plugin/types/Element.ts:35]: This line is producing a linter error, but the checks haven't run for this PR.

File: src/elements/image/imageElementDataTransformer.ts
Comment (SHession) [inline src/elements/image/imageElementDataTransformer.ts:63]: It's pleasing to be able to remove this, hopefully it still works as expected!

File: src/renderers/react/externalStore.ts
Comment (SHession) [inline src/renderers/react/externalStore.ts:8]: I think this is a really eloquent solution. As discussed when pairing, there may be a more fitting name.

File: src/renderers/react/externalStore.ts
Comment (SHession) [inline src/renderers/react/externalStore.ts:21]: Out of interest, what's the rationale / value behind this returning a boolean? Would a number of subscribers be as useful?

**Review by SHession — COMMENTED** [review]
Changes look sensible and I have added some comments for your consideration. 
I have had some issues running this locally, with what looks like an infinite loop when adding a new image in the sandbox.

File: src/renderers/react/externalStore.ts
Comment (jonathonherbert) [reply src/renderers/react/externalStore.ts:21]: We could also do that! It's to ensure that we're able to test unsubscription without having to dig in to implementation detail of the store.

File: src/plugin/types/Element.ts
Comment (jonathonherbert) [reply src/plugin/types/Element.ts:35]: This is not needed any more – we can remove.

File: src/elements/image/ImageElementForm.tsx
Comment (jonathonherbert) [reply src/elements/image/ImageElementForm.tsx:97]: Haha, yeah 😅

**Comment by jonathonherbert** [comment]
@SHession yes, there's a pathological case as a result of `none-selected` never appearing in our options list! Easy to resolve.

**Comment by jonathonherbert** [comment]
@SHession should now be resolved – I've tested with normal and thumbnail images on creation and crop. Interesting that we don't have tests for this (alongside our other elements, really!) – a blind spot we might want to remove.

File: demo/index.ts
Comment (SHession) [inline demo/index.ts:70]: Should the value of this item be `undefinedDropdownValue` to more closely reflect Composer?

File: src/elements/image/ImageElementForm.tsx
Comment (jonathonherbert) [reply src/elements/image/ImageElementForm.tsx:202]: Whups! Good spot. Surprised this isn't a linting error. I'll raise on https://github.com/guardian/configs/tree/main/packages/eslint-config-typescript.

**Comment by jonathonherbert** [comment]
@SHession that should now be fixed – worth noting that the first attempt at fixing the loop problem wasn't quite right.

Thinking that it was the logical result of a reference change in the `useEffect` dependencies, I memoised the options being passed to useEffect, but the problem persisted – creating a small image sent us into a loop where the image kept being destroyed and recreated.

It turns out that if we call `update` when the field is first created, it can be interfere with the Prosemirror lifecycle. Or at lest  that's my theory, which looks like:

 1. dispatch(1) starts: Image added
 2. Node created
 3. nodeView called
 4. React renders, and `useEffect` is triggered.
 5. Within `useEffect`, `field.update` is called, calling `updateOuterEditor` to update parent editor
 6. dispatch(2) starts in parent editor: but dispatch(1) not yet complete!

It's worth noting that I always thought `useEffect` was applied asynchronously, but that may not be the – React maintainers suggest that code run in effects [should be resilient to being run sync or async.](https://github.com/facebook/react/issues/22506#issuecomment-934542050).

So the solution is to defer the state updates for `CustomFieldView` by a tick or so to ensure they're run after Prosemirror is done with its state work. `CustomFieldView` is the only FieldView that exposes the opportunity to update state in this way, so it should be the only place we need to make `update()` async.

I've now tested by:

- creating a new image, with and without thumbnail
- updating image, with and without thumbnail
- randomising options and ensuring updates are correct

and all seems well.

File: src/plugin/fieldViews/CustomFieldView.ts
Comment (SHession) [inline src/plugin/fieldViews/CustomFieldView.ts:113]: This makes sense as a solution given your thorough explanation. I have also tested this fix against a bug I found in Composer and found it to be fixed.

**Review by SHession — APPROVED** [review]
@jonathonherbert, thanks for the thorough explanation. It explains the issue we saw with the Thumbnail option, and the solution should protect us from other potential problems in the future. 

I have tested this locally in the usual ways, I have also tested it against my ongoing work in Composer. It all appears to be working as expected and responding to changes nicely. 

Approving this now, but we should hold off merging until we are happy with the Composer changes.

**Review by SHession — APPROVED** [review]
Paired on these changes with @jonathonherbert after testing in the related Composer PR. I've Issues with adding a new element to a document displaying all dropdown options due to the Store not maintaining any state have now been resolved. The exports of certain values have also tidied up sone gaps in Composer. LGTM!

**Pull request merged**
